test/integration_test.py

```diff
@@ -214,36 +210,98 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
-        # No point in exhausting the iterator once we have a reasonable amount
-        # of inputs to randomly select from. This truncation prefers sources
-        # occurring first, so we shuffle them above to neutralize the bias.
-        bundle_fqids = islice(bundle_fqids, max_bundles * 10)
-        bundle_fqids = reservoir_sample(max_bundles, bundle_fqids, random=self.random)
-
-        if len(bundle_fqids) >= max_bundles:
-            log.warning('Not enough bundles in catalog %r. The test may fail.', catalog)
-
-        return bundle_fqids
+        sources = self.azul_client.catalog_sources(catalog)
+        # We need at least one managed_access bundle per IT. To index them with
+        # remote_reindex and avoid collateral bundles, we use as specific a
+        # a prefix as possible.
+        for source in self.managed_access_sources_by_catalog[catalog]:
+            assert str(source.spec) in sources
+            bundle_fqid = self.random.choice(
+                self.azul_client.list_bundles(catalog, source, prefix='')
+            )
+            # FIXME: We shouldn't need to include the common prefix
+            #        https://github.com/DataBiosphere/azul/issues/3579
+            common = source.spec.effective.prefix.common
+            prefix = bundle_fqid.uuid[len(common):8]
+            assert prefix != '', prefix
+            new_fqids = self.azul_client.list_bundles(catalog, source, prefix)
+            yield source, prefix, new_fqids
+
+    def _choose_prefix(self,
+                       catalog: CatalogName,
+                       source: SourceRef
+                       ) -> Tuple[str, List[SourcedBundleFQID]]:
+        fqids = self.azul_client.list_bundles(catalog, source, '')
+        fqids = self.azul_client.filter_obsolete_bundle_versions(fqids)
+        min_average_partition_size = 2
+        # Assuming an even distribution, this prefix length should yield at
+        # least min_average_partition_size bundles per prefix. The log base is
+        # 16 because prefixes are base 16.
+        prefix_len = math.floor(math.log(len(fqids) / min_average_partition_size + 1, 16))
+        if prefix_len == 0:
+            return '', fqids
+        else:
+            prefix = ''.join([
+                str(self.random.choice('abcdef0123456789'))
+                for _ in range(prefix_len)
+            ])
+            while True:
+                fqids = self.azul_client.list_bundles(catalog, source, prefix)
+                fqids = self.azul_client.filter_obsolete_bundle_versions(fqids)
+                # By the pigeonhole principle, this condition is guaranteed to
+                # be true eventually, and likely with few iterations
+                if len(fqids) > 0:
+                    return prefix, fqids
+                else:
+                    # Increment to the next prefix without carrying over
+                    prefix = hex(int(prefix, 16) + 1)[-prefix_len:]
+
+    def _choose_prefix_2(self,
```

Comment:
This is an earlier attempt which I don't like as much and will likely remove after Hannes reviews, assuming he prefers the other implementation.